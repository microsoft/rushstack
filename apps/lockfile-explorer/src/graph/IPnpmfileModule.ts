// Copyright (c) Microsoft Corporation. All rights reserved. Licensed under the MIT license.
// See LICENSE in the project root for license information.

import type { IPackageJson } from '@rushstack/node-core-library';

export interface IReadPackageContext {
  log: (message: string) => void;
}

export type IReadPackageHook = (
  packageJson: IPackageJson,
  context: IReadPackageContext
) => IPackageJson | Promise<IPackageJson>;

export interface IPnpmHooks {
  readPackage?: IReadPackageHook;
}

/**
 * Type of the `.pnpmfile.cjs` module.
 */
export interface IPnpmfileModule {
  hooks?: IPnpmHooks;
}
