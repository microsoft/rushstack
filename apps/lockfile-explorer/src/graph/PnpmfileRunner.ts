// Copyright (c) Microsoft Corporation. All rights reserved. Licensed under the MIT license.
// See LICENSE in the project root for license information.

import { Worker } from 'node:worker_threads';
import * as path from 'node:path';

import type { IPackageJson } from '@rushstack/node-core-library';

import type { IRequestMessage, ResponseMessage } from './pnpmfileRunnerWorkerThread';

interface IPromise {
  resolve: (r: IPackageJson) => void;
  reject: (e: Error) => void;
}

/**
 * Evals `.pnpmfile.cjs` in an isolated thread, so `transformPackageAsync()` can be used to rewrite
 * package.json files.  Calling `disposeAsync()` will free the loaded modules.
 */
export class PnpmfileRunner {
  private _worker: Worker;
  private _nextId: number = 1000;
  private _promisesById: Map<number, IPromise> = new Map();
  private _disposed: boolean = false;

  public logger: ((message: string) => void) | undefined = undefined;

  public constructor(pnpmfilePath: string) {
    this._worker = new Worker(path.join(`${__dirname}/pnpmfileRunnerWorkerThread.js`), {
      workerData: { pnpmfilePath }
    });

    this._worker.on('message', (message: ResponseMessage) => {
      const id: number = message.id;
      const promise: IPromise | undefined = this._promisesById.get(id);
      if (!promise) {
        return;
      }

      if (message.kind === 'return') {
        this._promisesById.delete(id);
        // TODO: Validate the user's readPackage() return value
        const result: IPackageJson = message.result as IPackageJson;
        promise.resolve(result);
      } else if (message.kind === 'log') {
        // No this._promisesById.delete(id) for this case
        if (this.logger) {
          this.logger(message.log);
        } else {
          console.log('.pnpmfile.cjs: ' + message.log);
        }
      } else {
        this._promisesById.delete(id);
        promise.reject(new Error(message.error || 'An unknown error occurred'));
      }
    });

    this._worker.on('error', (err) => {
      for (const promise of this._promisesById.values()) {
        promise.reject(err);
      }
      this._promisesById.clear();
    });

    this._worker.on('exit', (code) => {
      if (!this._disposed) {
        const error: Error = new Error(
          `PnpmfileRunner worker thread terminated unexpectedly with exit code ${code}`
        );
        console.error(error);
        for (const promise of this._promisesById.values()) {
          promise.reject(error);
        }
        this._promisesById.clear();
      }
    });
  }

  /**
   * Invokes the readPackage() hook from .pnpmfile.cjs
   */
  public transformPackageAsync(
    packageJson: IPackageJson,
    packageJsonFullPath: string
  ): Promise<IPackageJson> {
    if (this._disposed) {
      return Promise.reject(new Error('The operation failed because PnpmfileRunner has been disposed'));
    }

    const id: number = this._nextId++;
    return new Promise((resolve, reject) => {
      this._promisesById.set(id, { resolve, reject });
      this._worker.postMessage({ id, packageJson, packageJsonFullPath } satisfies IRequestMessage);
    });
  }

  public async disposeAsync(): Promise<void> {
    if (this._disposed) {
      return;
    }
    for (const pending of this._promisesById.values()) {
      pending.reject(new Error('Aborted because PnpmfileRunner was disposed'));
    }
    this._promisesById.clear();
    this._disposed = true;
    await this._worker.terminate();
  }
}
