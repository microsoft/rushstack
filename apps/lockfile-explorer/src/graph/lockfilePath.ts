// Copyright (c) Microsoft Corporation. All rights reserved. Licensed under the MIT license.
// See LICENSE in the project root for license information.

/**
 * For example, retrieves `d` from `/a/b/c/d`.
 */
export function getBaseNameOf(importerPath: string): string {
  if (importerPath.length === 0) {
    return '';
  }

  const index: number = importerPath.lastIndexOf('/');
  if (index === importerPath.length - 1) {
    throw new Error('Error: Path has a trailing slash');
  }
  if (index >= 0) {
    return importerPath.substring(index + 1);
  }
  return importerPath;
}

/**
 * For example, retrieves `/a/b/c` from `/a/b/c/d`.
 */
export function getParentOf(importerPath: string): string {
  if (importerPath === '' || importerPath === '.' || importerPath === '/') {
    throw new Error('Error: Path has no parent');
  }

  const index: number = importerPath.lastIndexOf('/');
  if (index === importerPath.length - 1) {
    throw new Error('Error: Path has a trailing slash');
  }
  if (index === 0) {
    return '/';
  }
  if (index < 0) {
    return '.';
  }
  return importerPath.substring(0, index);
}

/**
 * Cheaply resolves a relative path against a base path, assuming the paths are delimited by `/`,
 * and assuming the basePath is already in normal form.  An error occurs if the relative path
 * goes above the root folder.
 *
 * @example
 * ```ts
 * getAbsolutePath(`a/b/c`,  `d/e`)         === `a/b/c/d/e`
 * getAbsolutePath(`/a/b/c`, `d/e`)         === `/a/b/c/d/e`
 * getAbsolutePath(`/a/b/c`, `/d/e`)        === `/d/e`
 * getAbsolutePath(`a/b/c`,  `../../f`)     === `a/f`
 * getAbsolutePath(`a/b/c`,  `.././/f`)     === `a/b/f`
 * getAbsolutePath(`a/b/c`,  `../../..`)    === `.`
 * getAbsolutePath(`C:/a/b`, `../d`)        === `C:/a/d`
 * getAbsolutePath(`a/b/c`,  `../../../..`) === ERROR
 *
 * // Degenerate cases:
 * getAbsolutePath(`a/b/c/`, `d/`)          === `a/b/c/d`   // trailing slashes are discarded
 * getAbsolutePath(`./../c`, `d`)           === `./../c/d`  // basePath assumed to be normal form
 * getAbsolutePath(`C:\\`,   `\\a`)         === `C:\\/\\a`  // backslashes not supported
 * ```
 */
export function getAbsolute(basePath: string, relativePath: string): string {
  let leadingSlash: boolean;
  let stack: string[];

  // Discard intermediary slashes
  const relativeParts: string[] = relativePath.split('/').filter((part: string) => part.length > 0);
  if (relativePath.startsWith('/')) {
    stack = [];
    leadingSlash = true;
  } else {
    // Discard intermediary slashes
    stack = basePath.split('/').filter((part: string) => part.length > 0);
    leadingSlash = basePath.startsWith('/');
  }

  for (const part of relativeParts) {
    if (part === '.') {
      // current directory, do nothing
      continue;
    } else if (part === '..') {
      if (stack.length === 0) {
        throw new Error('getAbsolutePath(): relativePath goes above the root folder');
      }
      stack.pop();
    } else {
      stack.push(part);
    }
  }
  if (leadingSlash) {
    return '/' + stack.join('/');
  } else {
    return stack.length === 0 ? '.' : stack.join('/');
  }
}

/**
 * Returns the two parts joined by exactly one `/`, assuming the parts are already
 * in normalized form.  The `/` is not added if either part is an empty string.
 */
export function join(leftPart: string, rightPart: string): string {
  if (leftPart.length === 0) {
    return rightPart;
  }
  if (rightPart.length === 0) {
    return leftPart;
  }

  const leftEndsWithSlash: boolean = leftPart[leftPart.length - 1] === '/';
  const rightStartsWithSlash: boolean = rightPart[0] === '/';

  if (leftEndsWithSlash && rightStartsWithSlash) {
    return leftPart + rightPart.substring(1);
  }
  if (leftEndsWithSlash || rightStartsWithSlash) {
    return leftPart + rightPart;
  }
  return leftPart + '/' + rightPart;
}
