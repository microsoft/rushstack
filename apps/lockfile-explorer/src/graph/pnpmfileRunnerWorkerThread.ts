// Copyright (c) Microsoft Corporation. All rights reserved. Licensed under the MIT license.
// See LICENSE in the project root for license information.

import { parentPort, workerData, type MessagePort } from 'node:worker_threads';
import * as path from 'node:path';

import type { IPackageJson } from '@rushstack/node-core-library';

import type { IPnpmfileModule, IReadPackageContext } from './IPnpmfileModule';

export interface IRequestMessage {
  id: number;
  packageJson: IPackageJson;
  packageJsonFullPath: string;
}

export interface IResponseMessageLog {
  kind: 'log';
  id: number;
  log: string;
}
export interface IResponseMessageError {
  kind: 'error';
  id: number;
  error: string;
}
export interface IResponseMessageReturn {
  kind: 'return';
  id: number;
  result?: unknown;
}
export type ResponseMessage = IResponseMessageLog | IResponseMessageError | IResponseMessageReturn;

// debugger;

const { pnpmfilePath } = workerData;
const resolvedPath: string = path.resolve(pnpmfilePath);

let pnpmfileModule: IPnpmfileModule | undefined = undefined;
let pnpmfileModuleError: Error | undefined = undefined;

try {
  pnpmfileModule = require(resolvedPath);
} catch (error) {
  pnpmfileModuleError = error;
}

// eslint-disable-next-line @rushstack/no-new-null
const threadParentPort: null | MessagePort = parentPort;

if (!threadParentPort) {
  throw new Error('Not running in a worker thread');
}

threadParentPort.on('message', async (message: IRequestMessage) => {
  const { id, packageJson } = message;

  if (pnpmfileModuleError) {
    threadParentPort.postMessage({
      kind: 'error',
      id,
      error: pnpmfileModuleError.message
    } satisfies IResponseMessageError);
    return;
  }

  try {
    if (!pnpmfileModule || !pnpmfileModule.hooks || typeof pnpmfileModule.hooks.readPackage !== 'function') {
      // No transformation needed
      threadParentPort.postMessage({
        kind: 'return',
        id,
        result: packageJson
      } satisfies IResponseMessageReturn);
      return;
    }

    const pnpmContext: IReadPackageContext = {
      log: (logMessage) =>
        threadParentPort.postMessage({
          kind: 'log',
          id,
          log: logMessage
        } satisfies IResponseMessageLog)
    };

    const result: IPackageJson = await pnpmfileModule.hooks.readPackage({ ...packageJson }, pnpmContext);

    threadParentPort.postMessage({ kind: 'return', id, result } satisfies IResponseMessageReturn);
  } catch (e) {
    threadParentPort.postMessage({
      kind: 'error',
      id,
      error: (e as Error).message
    } satisfies IResponseMessageError);
  }
});
