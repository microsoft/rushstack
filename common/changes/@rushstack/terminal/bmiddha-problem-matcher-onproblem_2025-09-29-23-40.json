{
  "changes": [
    {
      "packageName": "@rushstack/terminal",
      "comment": "Add ProblemCollector.onProblem notification callback",
      "type": "minor"
    }
  ],
  "packageName": "@rushstack/terminal"
}