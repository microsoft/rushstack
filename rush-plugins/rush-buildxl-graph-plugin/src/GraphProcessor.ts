// Copyright (c) Microsoft Corporation. All rights reserved. Licensed under the MIT license.
// See LICENSE in the project root for license information.

import type { Operation, ILogger } from '@rushstack/rush-sdk';
import type { ShellOperationRunner } from '@rushstack/rush-sdk/lib/logic/operations/ShellOperationRunner';
import { Colorize } from '@rushstack/terminal';

import { filterObjectForDebug } from './debugGraphFiltering';

/**
 * @example
 * ```
 * {
 *   "id": "@rushstack/node-core-library#_phase:build",
 *   "task": "_phase:build",
 *   "package": "@rushstack/node-core-library",
 *   "dependencies": [
 *     "@rushstack/eslint-patch#_phase:build",
 *     "@rushstack/eslint-plugin#_phase:build",
 *     "@rushstack/eslint-plugin-packlets#_phase:build",
 *     "@rushstack/eslint-plugin-security#_phase:build"
 *   ],
 *   "workingDirectory": "/repo/libraries/node-core-library",
 *   "command": "heft run --only build -- --clean --production --drop-graph ./src/examples/graph.json"
 * }
 * ```
 *
 * See https://github.com/microsoft/BuildXL/blob/adf025c1b96b8106984928df3e9c30c8331bc8d6/Public/Src/Tools/JavaScript/Tool.RushGraphBuilder/src/RushBuildPluginGraph.ts
 *
 * @public
 */
export interface IGraphNode {
  /**
   * The unique id of the Pip
   *
   * @example
   * `@rushstack/node-core-library#_phase:build`
   */
  id: string;

  /**
   * The command to run during the Pip
   */
  command: string;

  /**
   * The working directory of the Pip
   */
  workingDirectory: string;

  /**
   * The project name associated with the Pip
   *
   * @example
   * `@rushstack/node-core-library`
   */
  package: string;

  /**
   * The task name of the Pip
   *
   * @example
   * `_phase:build`
   */
  task: string;

  /**
   * The IDs of the dependencies of the Pip. These are the {@link IGraphNode.id} properties of other Pips.
   */
  dependencies: string[];
}

interface IGraphNodeInternal extends Readonly<Omit<IGraphNode, 'dependencies' | 'command'>> {
  dependencies: ReadonlySet<string>;
  command: string | undefined;
}

type NodeMap = ReadonlyMap<string, IGraphNodeInternal>;

const REQUIRED_FIELDS: Array<keyof IGraphNodeInternal> = [
  // command is absent because it is not required
  'id',
  'task',
  'package',
  'dependencies',
  'workingDirectory'
];

/*
 * Try to get the operation id, return undefined if it fails
 */
export function tryGetOperationId(operation: Partial<Operation>): string | undefined {
  const task: string | undefined = operation.associatedPhase?.name;
  const project: string | undefined = operation.associatedProject?.packageName;
  return task && project ? `${project}#${task}` : undefined;
}

export class GraphProcessor {
  private readonly _logger: ILogger;

  public constructor(logger: ILogger) {
    this._logger = logger;
  }

  /*
   * Convert the operationMap into an array of graph nodes with empty commands pruned
   */
  public processOperations(operations: Set<Operation>): IGraphNode[] {
    const nodeMap: Map<string, IGraphNodeInternal> = new Map();
    for (const operation of operations) {
      const entry: IGraphNodeInternal = this._operationAsHashedEntry(operation);
      nodeMap.set(entry.id, entry);
    }

    return this._pruneNoOps(nodeMap);
  }

  /*
   * Validate that all dependencies exist
   * Validate that all nodes have a non-empty command
   * Print a message to the logger, and return true if the graph is valid
   */
  public validateGraph(entries: readonly Readonly<IGraphNode>[]): boolean {
    const entryIDs: Set<string> = new Set(entries.map((entry) => entry.id));
    let isValid: boolean = true;
    for (const entry of entries) {
      for (const depId of entry.dependencies) {
        if (!entryIDs.has(depId)) {
          this._logger.emitError(new Error(`${entry.id} has a dependency on ${depId} which does not exist`));
          isValid = false;
        }
      }

      if (!entry.command) {
        this._logger.emitError(new Error(`There is an empty command in ${entry.id}`));
        isValid = false;
      }
    }

    if (isValid) {
      this._logger.terminal.writeLine(
        Colorize.green('All nodes have non-empty commands and dependencies which exist')
      );
    }

    const totalEdges: number = entries.reduce((acc, entry) => acc + entry.dependencies.length, 0);
    this._logger.terminal.writeLine(`Graph has ${entries.length} nodes, ${totalEdges} edges`);
    return isValid;
  }

  /*
   * Get the operation id, throw an error if it fails
   */
  public getOperationId(operation: Operation): string {
    const result: string | undefined = tryGetOperationId(operation);
    if (!result) {
      throw new Error(
        `Operation does not have a name: ${JSON.stringify(filterObjectForDebug(operation, 2), undefined, 2)}`
      );
    }

    return result;
  }

  /*
   * remove all entries with empty commands
   * if an entry has a dependency with an empty command, it should inherit the dependencies of the empty command
   */
  private _pruneNoOps(inputNodeMap: NodeMap): IGraphNode[] {
    // Cache for the non-empty upstream dependencies of each operation
    const nonEmptyDependenciesByOperation: Map<IGraphNodeInternal, Set<string>> = new Map();
    function getNonEmptyDependencies(node: IGraphNodeInternal): ReadonlySet<string> {
      // If we've already computed this, return the cached result
      let nonEmptyDependencies: Set<string> | undefined = nonEmptyDependenciesByOperation.get(node);
      if (!nonEmptyDependencies) {
        nonEmptyDependencies = new Set();
        nonEmptyDependenciesByOperation.set(node, nonEmptyDependencies);
        for (const dependencyID of node.dependencies) {
          if (!inputNodeMap.get(dependencyID)!.command) {
            // If the dependency is empty, recursively inherit its non-empty dependencies
            for (const deepDependency of getNonEmptyDependencies(inputNodeMap.get(dependencyID)!)) {
              nonEmptyDependencies.add(deepDependency);
            }
          } else {
            nonEmptyDependencies.add(dependencyID);
          }
        }
      }

      return nonEmptyDependencies;
    }

    const result: IGraphNode[] = [];
    for (const node of inputNodeMap.values()) {
      const command: string | undefined = node.command;
      if (command) {
        const nonEmptyDependencySet: ReadonlySet<string> = getNonEmptyDependencies(node);
        result.push({
          ...node,
          dependencies: Array.from(nonEmptyDependencySet),
          command: command
        });
      }
    }

    return result;
  }

  /*
   * Convert an operation into a graph node
   */
  private _operationAsHashedEntry(operation: Operation): IGraphNodeInternal {
    const dependencies: Set<string> = new Set();
    for (const element of operation.dependencies.values()) {
      const id: string | undefined = this.getOperationId(element);
      if (id) {
        dependencies.add(id);
      }
    }

    const node: Partial<IGraphNodeInternal> = {
      id: tryGetOperationId(operation),
      task: operation.associatedPhase?.name,
      package: operation.associatedProject?.packageName,
      dependencies,
      workingDirectory: operation.associatedProject?.projectFolder,
      command: (operation.runner as ShellOperationRunner | undefined)?.commandToRun
    };

    const missingFields: (keyof IGraphNodeInternal)[] = [];
    for (const requiredField of REQUIRED_FIELDS) {
      if (!node[requiredField]) {
        missingFields.push(requiredField);
      }
    }

    if (missingFields.length > 0) {
      this._logger.emitError(
        new Error(`Operation is missing required fields ${missingFields.join(', ')}: ${JSON.stringify(node)}`)
      );
    }

    // the runner is a no-op if and only if the command is empty
    if (!!operation.runner?.isNoOp !== !node.command) {
      this._logger.emitError(
        new Error(`${node.id}: Operation runner isNoOp does not match commandToRun existence`)
      );
    }

    return node as IGraphNodeInternal;
  }
}
