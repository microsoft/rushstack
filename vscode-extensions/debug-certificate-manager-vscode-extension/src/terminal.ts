// Copyright (c) Microsoft Corporation. All rights reserved. Licensed under the MIT license.
// See LICENSE in the project root for license information.

import { ITerminal } from '@rushstack/terminal';
import * as vscode from 'vscode';
import { stripVTControlCharacters } from 'node:util';

export async function runWorkspaceCommandAsync({
  terminalOptions,
  commandLine,
  terminal
}: {
  terminalOptions: vscode.TerminalOptions;
  commandLine: string;
  terminal: ITerminal;
}): Promise<string> {
  const vsTerminal: vscode.Terminal = vscode.window.createTerminal(terminalOptions);

  // wait for shell to bootup and vs code shell integration to kick-in
  const shellIntegration: vscode.TerminalShellIntegration =
    vsTerminal.shellIntegration ??
    (await new Promise((resolve, reject) => {
      const shellIntegrationDisposable: vscode.Disposable = vscode.window.onDidChangeTerminalShellIntegration(
        (event) => {
          if (event.terminal !== vsTerminal) {
            return;
          }

          resolve(event.shellIntegration);
          shellIntegrationDisposable?.dispose();
        }
      );
    }));

  // Run the command through shell integration and grab output
  return new Promise<string>((resolve, reject) => {
    let outputStream: string = '';

    // start output capturing with the start execution event
    const startExecutionDisposable: vscode.Disposable = vscode.window.onDidStartTerminalShellExecution(
      async (event) => {
        if (event.terminal !== vsTerminal) {
          return;
        }
        terminal.writeLine(`Terminal shell execution started`);

        for await (const chunk of event.execution.read()) {
          outputStream += chunk;
        }
      }
    );

    // collect output and exit code
    const endExecutionDisposable: vscode.Disposable = vscode.window.onDidEndTerminalShellExecution(
      (event) => {
        if (event.terminal !== vsTerminal) {
          return;
        }

        terminal.writeLine(`Terminal shell execution ended with exit code ${event.exitCode}`);
        outputStream = outputStream.trim();
        outputStream = stripVTControlCharacters(outputStream);
        terminal.writeLine(`Terminal output: ${outputStream}`);

        endExecutionDisposable.dispose();
        startExecutionDisposable.dispose();

        if (event.exitCode === 0) {
          resolve(outputStream);
        } else {
          reject(outputStream);
        }
      }
    );

    shellIntegration.executeCommand(commandLine);
    terminal.writeLine(`Executing command: ${commandLine}`);
  }).finally(() => {
    vsTerminal.dispose();
  });
}
