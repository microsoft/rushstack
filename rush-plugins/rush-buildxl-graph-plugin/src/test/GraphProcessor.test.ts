// Copyright (c) Microsoft Corporation. All rights reserved. Licensed under the MIT license.
// See LICENSE in the project root for license information.

import type { IOperationRunner, Operation } from '@rushstack/rush-sdk';
import type { ShellOperationRunner } from '@rushstack/rush-sdk/lib/logic/operations/ShellOperationRunner';
import { Terminal, NoOpTerminalProvider } from '@rushstack/terminal';

import { GraphProcessor, type IGraphNode } from '../GraphProcessor';

// to update the examples folder, run the following command from the project root:
// export DEBUG_RUSH_BUILD_GRAPH="test" && rush build --production -t . --drop-graph ./src/examples/graph.json
import graph from '../examples/graph.json';
import debugGraph from '../examples/debug-graph.json';

function sortGraphNodes(graphNodes: IGraphNode[]): IGraphNode[] {
  return graphNodes.sort((a, b) => (a.id === b.id ? 0 : a.id < b.id ? -1 : 1));
}

describe(GraphProcessor.name, () => {
  let exampleGraph: readonly IGraphNode[];
  let graphParser: GraphProcessor;
  let emittedErrors: Error[];
  let emittedWarnings: Error[];

  beforeAll(() => {
    exampleGraph = sortGraphNodes(Array.from(graph.nodes));
  });

  beforeEach(() => {
    emittedErrors = [];
    emittedWarnings = [];

    const terminal: Terminal = new Terminal(new NoOpTerminalProvider());
    graphParser = new GraphProcessor({
      terminal,
      emitWarning: emittedWarnings.push.bind(emittedWarnings),
      emitError: emittedErrors.push.bind(emittedErrors)
    });
  });

  it('should process debug-graph.json into graph.json', () => {
    let prunedGraph: IGraphNode[] = graphParser.processOperations(
      new Set<Operation>(debugGraph.OperationMap as unknown as Operation[])
    );

    prunedGraph = sortGraphNodes(prunedGraph);
    expect(prunedGraph).toEqual(exampleGraph);
    expect(emittedErrors).toEqual([]);
    expect(emittedWarnings).toEqual([]);
  });

  it('should fail if the input schema is invalid', () => {
    const clonedOperationMap: Operation[] = JSON.parse(JSON.stringify(debugGraph.OperationMap));
    (clonedOperationMap[0].dependencies as unknown as Operation[]).push({
      incorrectPhase: { name: 'incorrectPhase' },
      incorrectProject: { packageName: 'incorrectProject' }
    } as unknown as Operation);
    const operations: Set<Operation> = new Set(clonedOperationMap);
    expect(() => graphParser.processOperations(operations)).toThrowErrorMatchingSnapshot();
    expect(emittedErrors).toEqual([]);
    expect(emittedWarnings).toEqual([]);
  });

  it('should fail if isNoOp mismatches a command', () => {
    const clonedOperationMap: Operation[] = JSON.parse(JSON.stringify(debugGraph.OperationMap));
    (clonedOperationMap[0].runner as IOperationRunner & { isNoOp: boolean }).isNoOp = true;
    (
      clonedOperationMap[0].runner as unknown as ShellOperationRunner & { commandToRun: string }
    ).commandToRun = 'echo "hello world"';
    const operations: Set<Operation> = new Set(clonedOperationMap);
    graphParser.processOperations(operations);
    expect(emittedErrors).not.toEqual([]);
    expect(emittedWarnings).toEqual([]);
  });

  describe(GraphProcessor.prototype.validateGraph.name, () => {
    it('should validate graph.json', () => {
      const isValid: boolean = graphParser.validateGraph(exampleGraph);
      expect(isValid).toBe(true);
      expect(emittedErrors).toEqual([]);
      expect(emittedWarnings).toEqual([]);
    });

    it('should fail to validate when command is empty', () => {
      const invalidGraph = JSON.parse(JSON.stringify(exampleGraph));
      invalidGraph[2].command = '';
      const isValid: boolean = graphParser.validateGraph(invalidGraph);
      expect(isValid).toBe(false);
      expect(emittedErrors).toMatchSnapshot();
      expect(emittedWarnings).toEqual([]);
    });

    it('should fail to validate when command is missing', () => {
      const invalidGraph = JSON.parse(JSON.stringify(exampleGraph));
      delete invalidGraph[2].command;
      const isValid: boolean = graphParser.validateGraph(invalidGraph);
      expect(isValid).toBe(false);
      expect(emittedErrors).toMatchSnapshot();
      expect(emittedWarnings).toEqual([]);
    });

    it('should fail to validate when id is missing', () => {
      const invalidGraph = JSON.parse(JSON.stringify(exampleGraph));
      delete invalidGraph[2].id;
      const isValid: boolean = graphParser.validateGraph(invalidGraph);
      expect(isValid).toBe(false);
      expect(emittedErrors).toMatchSnapshot();
      expect(emittedWarnings).toEqual([]);
    });

    it('should fail to validate when id is empty', () => {
      const invalidGraph = JSON.parse(JSON.stringify(exampleGraph));
      invalidGraph[2].id = '';
      const isValid: boolean = graphParser.validateGraph(invalidGraph);
      expect(isValid).toBe(false);
      expect(emittedErrors).toMatchSnapshot();
      expect(emittedWarnings).toEqual([]);
    });

    it('should fail to validate when dependencies refer to nonexistent nodes', () => {
      const invalidGraph = JSON.parse(JSON.stringify(exampleGraph));
      invalidGraph[2].dependencies.push('@this/is/presumably#_not:real');
      const isValid: boolean = graphParser.validateGraph(invalidGraph);
      expect(isValid).toBe(false);
      expect(emittedErrors).toMatchSnapshot();
      expect(emittedWarnings).toEqual([]);
    });
  });
});
