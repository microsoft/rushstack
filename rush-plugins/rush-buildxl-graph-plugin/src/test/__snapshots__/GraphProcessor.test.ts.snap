// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`GraphProcessor should fail if the input schema is invalid 1`] = `
"Operation does not have a name: {
  \\"incorrectPhase\\": {
    \\"name\\": \\"incorrectPhase\\"
  },
  \\"incorrectProject\\": {
    \\"packageName\\": \\"incorrectProject\\"
  }
}"
`;

exports[`GraphProcessor validateGraph should fail to validate when command is empty 1`] = `
Array [
  [Error: There is an empty command in @microsoft/rush-lib#_phase:build],
]
`;

exports[`GraphProcessor validateGraph should fail to validate when command is missing 1`] = `
Array [
  [Error: There is an empty command in @microsoft/rush-lib#_phase:build],
]
`;

exports[`GraphProcessor validateGraph should fail to validate when dependencies refer to nonexistent nodes 1`] = `
Array [
  [Error: @microsoft/rush-lib#_phase:build has a dependency on @this/is/presumably#_not:real which does not exist],
]
`;

exports[`GraphProcessor validateGraph should fail to validate when id is empty 1`] = `
Array [
  [Error: @rushstack/rush-buildxl-graph-plugin#_phase:build has a dependency on @microsoft/rush-lib#_phase:build which does not exist],
  [Error: @rushstack/rush-sdk#_phase:build has a dependency on @microsoft/rush-lib#_phase:build which does not exist],
]
`;

exports[`GraphProcessor validateGraph should fail to validate when id is missing 1`] = `
Array [
  [Error: @rushstack/rush-buildxl-graph-plugin#_phase:build has a dependency on @microsoft/rush-lib#_phase:build which does not exist],
  [Error: @rushstack/rush-sdk#_phase:build has a dependency on @microsoft/rush-lib#_phase:build which does not exist],
]
`;
